### DISCUSSION TITLE
Many bugs and malfunctions lately, what's going on?

### DISCUSSION THREAD

**Discussion by not-available-username** [mozilla-mobile/firefox-ios#16156] [General]
As requested, I'm posting here to follow up on this #16152, as I'm wondering if something else is happening with the development of this application. As the months go by, I'm noticing an increasingly poor user experience, with lots of bugs, unresponsive pages and freezes. I know this isn't helpful in tracking the issues (and frankly I'm not very keen, a minor bug I reported that is still there has been marked as fixed).

I'm curious to know if this is due to a reduction in active developers, or some major changes by Meta, Google and Apple (as they are ubiquitous on the web and not happy with Firefox's privacy, I guess), or something else. 

I use Firefox on all my devices (iOS, MacOs and Windows) and I want to avoid switching to something else on iOS.

**Comment by lmarceau** [comment]
I'll try to explain as best as I can, starting with a bit of history of the project. Firefox for iOS has been around for a while, being created around 2015 I believe. At that time, Swift (the programming language we develop in within this project) was fairly new, and a lot of the patterns that were used at that time didn't age so well. Adding on top of that, the project was left in maintenance mode for a certain period of time. Those two things combined brings us to today - the application being a tangled mess. We are a team of 9 devs both developing and adding new features in as the company needs/while also working on untangling the spaghetti code that has been living for a long time. Untangling the code is a very risky business, which is the cause of some of the problems we saw in the last months. But continuing to add features on top of the spaghetti without untangling it is unsustainable as well. So we're walking a fine thread, de-risking changes as much as we can, but sadly some of the risks were felt recently by users. We're actively working on improving the application, but this improvement means we might see instability. We have our team of QA, testers, unit tests, UI tests, etc - but things can still slip, particularly when we're working with a codebase like this one crippled with tech debt. I hope this answer helps to understand a bit of the challenges we face. It doesn't excuse in any way problems that are currently live in the app, but at least having some explanation can (I hope) help understand a bit? Thanks for reading, let me know if you have any further questions/comments!

**Reply by not-available-username** [reply]
Hi! Thank you for the detailed explanation, now I understand. 

Unfortunately the vast majority of the users don't even know that Firefox has this type of development and how open source works, I really hope that you can gather more resources and improve Firefox for iOS, thank you!
